jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/BufferingResponseListener.java

```diff
@@ -177,18 +180,31 @@ public String getContentAsString(String encoding)
      */
     public String getContentAsString(Charset encoding)
     {
-        if (buffer == null)
+        if (buffer.getLength() == 0)
             return null;
-        return BufferUtil.toString(buffer, encoding);
+        return new String(toByteArray(), encoding);
     }
 
     /**
      * @return Content as InputStream
      */
     public InputStream getContentAsInputStream()
     {
-        if (buffer == null)
-            return new ByteArrayInputStream(new byte[0]);
-        return new ByteArrayInputStream(buffer.array(), buffer.arrayOffset(), buffer.remaining());
+        return new ByteArrayInputStream(toByteArray());
+    }
+
+    private byte[] toByteArray()
+    {
+        synchronized (buffer)
+        {
+            try (ByteBufferAccumulator buffer = this.buffer)
+            {
+                if (buffer.getLength() == 0)
+                    result = new byte[0];
+                else
+                    result = buffer.toByteArray();
+            }
+        }
+        return result;
```

Comment:
ideally we would not have to copy all the retained buffers to a single buffer just to stream as a InputStream.  We should have a mechanism to do this without this extra copy.